code/parachain/frame/composable-traits/src/ibc.rs

```diff
@@ -0,0 +1,145 @@
+use serde_json::Value;
+
+use crate::prelude::*;
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
+#[serde(rename_all = "snake_case")]
+#[cfg_attr(feature = "std", derive(JsonSchema))]
+pub enum IBCLifecycleComplete {
+    #[serde(rename = "ibc_ack")]
+    IBCAck {
+        /// The source channel (osmosis side) of the IBC packet
+        channel: String,
+        /// The sequence number that the packet was sent with
+        sequence: u64,
+        /// String encoded version of the ack as seen by OnAcknowledgementPacket(..)
+        ack: String,
+        /// Weather an ack is a success of failure according to the transfer spec
+        success: bool,
+    },
+    #[serde(rename = "ibc_timeout")]
+    IBCTimeout {
+        /// The source channel (osmosis side) of the IBC packet
+        channel: String,
+        /// The sequence number that the packet was sent with
+        sequence: u64,
+    },
+}
+
+/// This message should be send as part of wasm termination memo.
+/// So that can match it to sender hash and know what channel and origin was used to send message.
+/// All information here is not secured until compared with existing secured data.
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
+#[serde(rename_all = "snake_case")]
+#[cfg_attr(feature = "std", derive(JsonSchema))]
+pub struct VerifiableWasmMsg {
+    pub bech32_prefix : String,
+    pub channel : String,
+    pub original_sender : String,
+    pub asset : Coin,
+}
+
+/// Message type for `sudo` entry_point
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
+#[serde(rename_all = "snake_case")]
+#[cfg_attr(feature = "std", derive(JsonSchema))]
+pub enum SudoMsg {
+    #[serde(rename = "ibc_lifecycle_complete")]
+    IBCLifecycleComplete(IBCLifecycleComplete),
+}
+
+pub const SENDER_PREFIX : &str = "ibc-wasm-hook-intermediary";
+
+/// from Go code to make compliant wasm hook
+pub fn derive_intermediate_sender(channel: String, original_sender : String, bech32_prefix : String) -> Result<String, ()> {
```

Comment:
It’s more idiomatic to take the arguments by reference:

```suggestion
pub fn derive_intermediate_sender(channel: &str, original_sender: &str, bech32_prefix: &str) -> Result<String, ()> {
```

This way caller is not forced to make an owned object which this
function discards anyway.

Reply:
yeah, sorry. just direct cp paste from go. did not noticed